### PR TITLE
[CBRD-21433] Enhancement for loading a foreign key

### PR DESCRIPTION
**Pull request by lucianlazar2211**
http://jira.cubrid.org/browse/CBRD-21433

This enhancement provide a better loading time of the foreign key.  In the previous iteration, each value from the foreign key was checked to reside also in the primary key, by scanning the primary key tree always from the root to leafs. Now, just the first value from the key will be scanned from root to leafs, while the others will be looked for starting from the leaf where the last value resides.

This is an early version of it. I have opened the PR for easier discussion over it. Things that need to be implemented:

- [x]  Check for descending type of index and treat it accordingly

- [x]  Use a range_scan_advance_over_filtered_keys type routine for the values in the foreign key

Testing status:

- SQL regressions -> Completed
- Medium regressions -> Completed
- CTL regressions -> In progress

### CODE REVIEW

File: src/storage/btree_load.c [hunk 1]
```diff
@@ -3878,3 +3875,385 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
...
+	  pk_header = btree_get_node_header (thread_p, pk_leaf);
+
+	  /* Get the number of keys in current leaf.  */
+	  pk_key_cnt = btree_node_number_of_keys (thread_p, pk_leaf);
+
+	  /* We try to resume the search in the current leaf. */
+	  while (!found_p && slot_id <= pk_key_cnt)
+	    {
+	      /* Try to get the value from the next slot. */
+	      ret = get_value_from_leaf_slot (thread_p, &(pk_bt_scan.btid_int), pk_leaf, slot_id, &pk_key);
```
Comment (ribram) [inline src/storage/btree_load.c]: use advance_to_next_slot_and_fix_page to avoid duplicating the code.

File: src/storage/btree_load.c [hunk 2]
```diff
@@ -3878,3 +3875,385 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
...
+
+	      if (ret != NO_ERROR)
+		{
+		  goto end;
+		}
+
+	      /* We need to compare the current value with the new value from the primary key. */
+	      ret = btree_compare_key (&pk_key, &fk_key, pk_bt_scan.btid_int.key_type, 1, 1, NULL);
```
Comment (ribram) [inline src/storage/btree_load.c]: Declare "DB_VALUE_COMPARE_RESULT compare_ret;" at the beginning of function and here do:
```c
	      /* We need to compare the current value with the new value from the primary key. */
	      compare_ret = btree_compare_key (&pk_key, &fk_key, pk_bt_scan.btid_int.key_type, 1, 1, NULL);
              if (compare_ret == DB_EQ)
                {
                  /* found, stop searching in pk */
                  break;
                }
              else if (compare_ret == DB_LT)
                {
                  /* continue looking */
                  continue;
                }
              else
                {
                  /* set fk_invalid error */
                  /* goto end */
                }
```

File: src/storage/btree.c [hunk 3]
```diff
@@ -24207,7 +24183,7 @@ btree_range_scan_read_record (THREAD_ENTRY * thread_p, BTREE_SCAN * bts)
  * thread_p (in)  : Thread entry.
  * bts (in/out)	  : B-tree scan helper.
  */
-static int
+int
```
Comment (ribram) [inline src/storage/btree.c]: this change can be rollbacked.

File: src/storage/btree.c [hunk 4]
```diff
@@ -14054,14 +14051,14 @@ btree_locate_key (THREAD_ENTRY * thread_p, BTID_INT * btid_int, DB_VALUE * key,
...
-					    btree_advance_and_find_key, slot_id, NULL, NULL, &search_key,
-					    &leaf_page) != NO_ERROR)
+  error = btree_search_key_and_apply_functions (thread_p, btid_int->sys_btid, NULL, key, NULL, NULL,
+						btree_advance_and_find_key, slot_id, NULL, NULL, &search_key,
+						&leaf_page);
+  if (error != NO_ERROR)
     {
-      /* Error */
-      ASSERT_ERROR ();
       assert (leaf_page == NULL);
```
Comment (ribram) [inline src/storage/btree.c]: Also add ASSERT_ERROR ();

File: src/storage/btree.c [hunk 5]
```diff
@@ -14035,11 +14030,13 @@ btree_reflect_global_unique_statistics (THREAD_ENTRY * thread_p, GLOBAL_UNIQUE_S
  * Note: Search the B+tree index to locate the page and record that contains
  *	 the key, or would contain the key if the key was to be located.
  */
-static PAGE_PTR
+int
```
Comment (ribram) [inline src/storage/btree.c]: Reflect signature changes in comment.

File: src/storage/btree_load.c [hunk 6]
```diff
@@ -895,6 +905,13 @@ xbtree_load_index (THREAD_ENTRY * thread_p, BTID * btid, const char *bt_name, TP
       /* No need to deal with overflow pages anymore */
       load_args->ovf.pgptr = NULL;
 
+      /* Check the correctness of the foreign key, if any. */
+      if (has_fk)
+	{
+	  if (btree_load_check_fk (thread_p, load_args, sort_args) != NO_ERROR)
+	    goto error;
```
Comment (ribram) [inline src/storage/btree_load.c]: > goto error;

please use brackets:
```c
	  {
	    goto error;
	  }
```

File: src/storage/btree_load.c [hunk 7]
```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
     }
   return PERF_PAGE_BTREE_ROOT;
 }
+
+/*
+ * btree_load_check_fk () -
```
Comment (ribram) [inline src/storage/btree_load.c]: Update comment & signature:

1. add short function description
2. remove _local suffix from load_args and sort_args.
3. loag_args: context for loaded index (includes leaf level of foreign key)
    sort_args: context for sorting (includes info on primary key)

File: src/storage/btree_load.c [hunk 8]
```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
...
+ *
+ *    return:				- NO_ERROR or error code.
+ *    load_args_local(in):		- Information about the foreign key
+ *    sort_args_local(in):		- Information about the primary key
+ *
+ */
+int
+btree_load_check_fk (THREAD_ENTRY * thread_p, const LOAD_ARGS * load_args, const SORT_ARGS * sort_args)
+{
+  DB_VALUE fk_key, pk_key;
```
Comment (ribram) [inline src/storage/btree_load.c]: Rename for clarity: for fk/pk_key_cnt and fk_pk_header add node_ inside name (e.g. fk_node_key_cnt, fk_node_header). They may be confused for index total count and header.

Remove unused advance_direction.

File: src/storage/btree_load.c [hunk 9]
```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
...
+    {
+      pgbuf_unfix_and_init (thread_p, curr_fk_pageptr);
+    }
+
+  if (pk_bt_scan.C_page)
+    {
+      pgbuf_unfix_and_init (thread_p, pk_bt_scan.C_page);
+    }
+
+  if (load_args->leaf.pgptr)
```
Comment (ribram) [inline src/storage/btree_load.c]: do not unfix load_args->leaf.pgptr. it was not fixed here.

free val_print.

File: src/storage/btree_load.c [hunk 10]
```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
...
+      /* Get the last leaf. Current leaf is the last one. */
+      vpid = load_args->leaf.vpid;
+    }
+
+  /* Init slot id */
+  pk_bt_scan.slot_id = 0;
+
+  while (true)
+    {
+      ret = btree_advance_to_next_slot_and_fix_page (thread_p, sort_args->btid, &vpid, &curr_fk_pageptr, &fk_slot_id,
```
Comment (ribram) [inline src/storage/btree_load.c]: foreign key should be scanned in is_fk_scan_desc.

What we want is to scan keys in the same order as they are in primary key. Nobody else accesses newly loaded foreign key, so it does not matter in which order we'll scan it. However, concurrent range scans can be executed on primary key, so it is better to scan it in ascending order.

Therefore, if primary keys are ordered descending, while foreign keys are ordered ascending, we need to scan foreign key in descending order. is_fk_scan_desc reflects the right scan order.

File: src/storage/btree_load.c [hunk 11]
```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
...
+    {
+      ASSERT_ERROR ();
+      return ret;
+    }
+
+  return ret;
+}
+
+/*
+ * btree_advance_to_next_slot_and_fix_page () -
```
Comment (ribram) [inline src/storage/btree_load.c]: Add function description mentioning that it advances from one page to another without conditional latches. Warn callers that it should not used for descending scans if no concurrency cannot be guaranteed.

File: src/storage/btree_load.c [hunk 12]
```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
...
+  PAGE_PTR old_page = NULL;
+
+  /* Clear current key, if any. */
+  if (!DB_IS_NULL (key))
+    {
+      pr_clear_value (key);
+    }
+
+start_alg:
+  if (page != NULL)
```
Comment (ribram) [inline src/storage/btree_load.c]: The function should handle page change. Also, it's better to avoid goto labels.

do something like:

```

if not initialized (current page is nil)
  current page  = fix page
  check error
  initialize slot, node key count and node header

/* advance to next key (having at least one visible object) */
while (true) {
  increment slot;
  if slot not in page
    if no next/prev page
      unfix current page
      break loop
    else
      next page = fix next/prev page
      check error
      unfix current page
      current page = next page
      initialize slot, node key count and node header
  if snapshot != NULL
    count visible objects
    check error
    if visible == 0
      continue
  /* key is valid */
  break
}
```

File: src/storage/btree_load.c [hunk 13]
```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
...
+
+  if (ret != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      return ret;
+    }
+
+
+  num_visible =
+    btree_get_num_visible_from_leaf_and_ovf (thread_p, btid, &record, key_offset, &leaf, NULL, mvcc_snapshot);
```
Comment (ribram) [inline src/storage/btree_load.c]: please update btree_get_num_visible_from_leaf_and_ovf and btree_record_get_num_visible_oids to return error and output number of visible as argument.

File: src/storage/btree_load.c [hunk 14]
```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
...
+ *  thread_p(in):		    -	Thread entry.
+ *  btid(in):			    -	B-tree info.
+ *  pg_ptr(in):			    -	Page pointer.
+ *  mvcc_snapshot(in):		    -	The MVCC snapshot.
+ *
+ *  return: Number of visible items, or error code if any error occurs.	    
+ */
+
+static int
+btree_get_number_of_items (THREAD_ENTRY * thread_p, BTID_INT * btid, PAGE_PTR pg_ptr, MVCC_SNAPSHOT * mvcc_snapshot)
```
Comment (ribram) [inline src/storage/btree_load.c]: convert btree_get_number_of_items to btree_has_any_visible. return error code and output true/false. if snapshot is nil, early out with answer true.

File: src/storage/btree_load.c [hunk 15]
```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
...
+    }
+
+  /* Init slot id */
+  pk_bt_scan.slot_id = 0;
+
+  while (true)
+    {
+      ret = btree_advance_to_next_slot_and_fix_page (thread_p, sort_args->btid, &vpid, &curr_fk_pageptr, &fk_slot_id,
+						     &fk_key, sort_args->key_type->is_desc, &fk_key_cnt, &fk_header,
+						     &mvcc_snapshot_dirty);
```
Comment (ribram) [inline src/storage/btree_load.c]: I think I told you to add mvcc_snapshot_dirty here, but I was wrong. btree_sort_get_next already checks dirty snapshot, therefore only objects passing the checks are loaded. all keys will pass the visibility check in foreign key.

However, the dirty snapshot is required to check primary keys.

File: src/storage/btree_load.c [hunk 16]
```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
...
+	{
+	  /* Only way to get this is by having no visible objects in the foreign key. */
+	  /* Must be checked!! */
+	  break;
+	}
+
+      /* We got the value from the foreign key, now search through the primary key index. */
+      found = false;
+
+    start_primary_search:
```
Comment (ribram) [inline src/storage/btree_load.c]: advancing to new page should be handled by btree_advance_to_next_slot_and_fix_page and start_primary_search should be removed.

File: src/storage/btree_load.c [hunk 17]
```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
...
+	      assert (pk_bt_scan.C_page != NULL);
+	    }
+
+	  /* Unfix old page, if any. */
+	  if (old_page != NULL)
+	    {
+	      pgbuf_unfix_and_init (thread_p, old_page);
+	    }
+
+	  /* Value was found, proceed with the next value */
```
Comment (ribram) [inline src/storage/btree_load.c]: first primary key must be checked for visible objects too.

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/storage/btree_load.c (166 incorrect indentation(s) found)
```
  3832 						     &fk_key, is_fk_scan_desc, &fk_node_key_cnt, &fk_node_header, NULL);
  3834 	{
  3835 	  ASSERT_ERROR ();
  3836 	  break;
  3837 	}
  3840 	{
  3841 	  /* Search has ended. */
  3842 	  break;
  3843 	}
  3846 	{
  3847 	  /* Only way to get this is by having no visible objects in the foreign key. */
  3848 	  /* Must be checked!! */
  3849 	  break;
  3850 	}
  3857 	{
  3858 	  /* No search has been initiated yet, we start from root. */
  3859 	  ret =
  3860 	    btree_locate_key (thread_p, &pk_bt_scan.btid_int, &fk_key, &pk_bt_scan.C_vpid, &pk_bt_scan.slot_id,
  3861 			      &pk_bt_scan.C_page, &found);
  3862 
```
... More than 20 lines are omitted.
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by lucianlazar2211** [comment]
While I was working on this issue I spotted an unusual behavior regarding the creation of the foreign key, on the develop/release branch.

Referencing an entry from the CUBRID manual regarding foreign keys that can be found here : [Foreign Key](http://www.cubrid.org/manual/en/10.1/sql/schema/table_stmt.html#foreign-key-constraint) .

`For each row R1 of the referencing table, there should be some row R2 of the referenced table such that the value of each referencing column in R1 is either NULL or is equal to the value of the corresponding referenced column in R2.`

On a basic database, use the next scenario:
```
drop if exists t2, t1;
create table t1 (id int not null auto_increment, parent_id int default 1, primary key (id,parent_id));
create table t2(id int auto_increment,i int,j int);
alter table t1 partition by hash (id) partitions 2;
alter table t2 partition by hash (id) partitions 4;
insert into t1 values(NULL,1),(NULL,2),(NULL,2);
insert into t2 values(4,NULL,1);
```
The two tables will look like this:
```
select * from t1;

=== <Result of SELECT Command in Line 1> ===

           id    parent_id
==========================
            3            2
            1            1
            2            2

3 rows selected. (0.017823 sec) Committed.

select * from t2;

=== <Result of SELECT Command in Line 1> ===

           id            i            j
=======================================
            4         NULL            1
```

Now we try to create a foreign key on t2 referencing t1 using the next query:
```
alter table t2 add constraint test_ibfk_1 foreign key (id,i) references t1 (id,parent_id);
Execute OK. (3.133985 sec) Committed.
```

The foreign key was successfully created, however the column `id` from `t2` has a value of `4`, that is not referenced into the `id` column in `t1`. Judging by the reference in the CUBRID manual, I think this might be incorrect.

Is this the intended behavior though?

On a side note, by checking the following code:
```
btree_load.c:2884:

 if (n_attrs > 1)
    {
      has_null = btree_multicol_key_has_null (keyval);
    }
  else
    {
      has_null = DB_IS_NULL (keyval);
    }

```

I think there should be used `btree_multicol_key_is_null (keyval)` instead.

**Comment by eseokoh** [comment]
@lucianlazar2211 

Yes, it is intended.

ANSI SQL says 

> The choices for <match type> are MATCH SIMPLE, MATCH PARTIAL, and MATCH FULL; MATCH
SIMPLE is the default. There is no semantic difference between these choices if there is only one referencing column (and, hence, only one referenced column). There is also no semantic difference if all referencing columns are not nullable. If there is more than one referencing column, at least one of which is nullable, and if no <referencing period specification> is specified, then the various <match type>s have the following semantics:
— MATCH SIMPLE: if at least one referencing column is null, then the row of the referencing table passes the constraint check. If all referencing columns are not null, then the row passes the constraint check if and only if there is a row of the referenced table that matches all the referencing columns.
— MATCH PARTIAL: if all referencing columns are null, then the row of the referencing table passes the constraint check. If at least one referencing columns is not null, then the row passes the constraint check if and only if there is a row of the referenced table that matches all the non-null referencing columns.
— MATCH FULL: if all referencing columns are null, then the row of the referencing table passes the constraint check. If all referencing columns are not null, then the row passes the constraint check if and only if there is a row of the referenced table that matches all the referencing columns. If some referencing column is null and another referencing column is nonnull, then the row of the referencing table violates the constraint
check.

In short, we don't provide options for <match type> and our behavior is <MATCH SIMPLE> which is the default behavior of ANSI SQL and the other (commercial) products.

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/query/partition.c (1 incorrect indentation(s) found)
```
  3889 				 int *position)
```
#### File: src/query/partition.h (1 incorrect indentation(s) found)
```
   124 					    BTID * btid, int *position);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------
